File: ports/ruby/lib/unicode_ruby/security/form/normalization_bomb.rb
# frozen_string_literal: true

require_relative "../identity/ucd"

module UnicodeRuby
  module Security
    module Form
      # Normalization-bomb detection — inputs whose NFD or NFKD expansion
      # exceeds documented bounds, the classic normalization-expansion DoS.
      # Priority: a per-codepoint blow-up scan, an overall NFKD ratio, an
      # overall NFD ratio.  Ratios are expressed in hundredths to avoid floats.
      module NormalizationBomb
        # Maximum allowed NFKD expansion per single codepoint.
        MAX_NFKD_PER_CP = 8
        # Overall NFD expansion ratio threshold, in hundredths (300 = 3x).
        NFD_RATIO_PCT = 300
        # Overall NFKD expansion ratio threshold, in hundredths (400 = 4x).
        NFKD_RATIO_PCT = 400

        Detection = Struct.new(:sub, :positions)

        module_function

        # First position whose single-codepoint NFKD expansion exceeds the cap.
        def first_blowup_cp(input)
          input.each_index do |index|
            expand = Ucd.to_nfkd([input[index]]).length
            return index if expand > MAX_NFKD_PER_CP
          end
          nil
        end

        def nfd_ratio_pct(input)
          return 0 if input.empty?

          Ucd.to_nfd(input).length * 100 / input.length
        end

        def nfkd_ratio_pct(input)
          return 0 if input.empty?

          Ucd.to_nfkd(input).length * 100 / input.length
        end

        def detect(input)
          pos = first_blowup_cp(input)
          return Detection.new("SingleCpBlowup", [pos]) unless pos.nil?

          return Detection.new("NfkdHighExpansion", []) if nfkd_ratio_pct(input) > NFKD_RATIO_PCT
          return Detection.new("NfdHighExpansion", []) if nfd_ratio_pct(input) > NFD_RATIO_PCT

          Detection.new(nil, [])
        end
      end
    end
  end
end

File: ports/python/src/unicode_python/security/form/normalization_bomb.py
"""Normalization-bomb detector (F1).

Mirrors ``Unicode.Security.Form.NormalizationBomb``. Detects inputs whose NFD
or NFKD expansion exceeds documented bounds — the classic
normalization-expansion DoS, where a small input expands to a very large
normalized form and exhausts memory/CPU at the receiving layer (the Arabic
ligature U+FDFA expands to 18 codepoints under NFKD, etc.).

Pure functional: compute NFD and NFKD lengths, then three priority-ordered
checks — a per-codepoint blow-up scan, an overall NFKD ratio, an overall NFD
ratio. Ratios are expressed in hundredths to avoid floats.
"""

from dataclasses import dataclass

from ..identity.ucd import to_nfd, to_nfkd

__all__ = ["Detection", "detect"]

# Maximum allowed NFKD expansion per single codepoint. Hangul <= 3, Greek
# extended forms 4, the largest non-FDFA Arabic ligature (FDFB) 8; anything
# greater than 8 is flagged.
MAX_NFKD_PER_CP = 8

# Overall-sequence NFD expansion ratio threshold, in hundredths (300 = 3x).
# Pure Hangul sits at exactly 300 and stays clear under strict ``>``.
NFD_RATIO_PCT = 300

# Overall-sequence NFKD expansion ratio threshold, in hundredths (400 = 4x).
NFKD_RATIO_PCT = 400


@dataclass(frozen=True, slots=True)
class Detection:
    """One normalization-bomb scan result. ``sub`` is ``None`` for a clear
    input; a per-codepoint blow-up carries the offending position, the ratio
    hazards carry no position."""

    sub: str | None
    positions: tuple[int, ...]


def _first_blowup_cp(input_cps: list[int]) -> tuple[int, int, int] | None:
    """First position whose single-codepoint NFKD expansion exceeds
    ``MAX_NFKD_PER_CP``, with the codepoint and its expansion length."""
    for index, cp in enumerate(input_cps):
        expand = len(to_nfkd([cp]))
        if expand > MAX_NFKD_PER_CP:
            return (index, cp, expand)
    return None


def _nfd_ratio_pct(input_cps: list[int]) -> int:
    """NFD ratio percentage (``100 * nfdLen // inputLen``); 0 on empty input."""
    if not input_cps:
        return 0
    return len(to_nfd(input_cps)) * 100 // len(input_cps)


def _nfkd_ratio_pct(input_cps: list[int]) -> int:
    """NFKD ratio percentage (``100 * nfkdLen // inputLen``); 0 on empty input."""
    if not input_cps:
        return 0
    return len(to_nfkd(input_cps)) * 100 // len(input_cps)


def detect(input_cps: list[int]) -> Detection:
    """Detect a normalization-expansion bomb. Priority: per-codepoint blow-up,
    then overall NFKD ratio, then overall NFD ratio."""
    blowup = _first_blowup_cp(input_cps)
    if blowup is not None:
        pos, _cp, _expand = blowup
        return Detection(sub="SingleCpBlowup", positions=(pos,))
    if _nfkd_ratio_pct(input_cps) > NFKD_RATIO_PCT:
        return Detection(sub="NfkdHighExpansion", positions=())
    if _nfd_ratio_pct(input_cps) > NFD_RATIO_PCT:
        return Detection(sub="NfdHighExpansion", positions=())
    return Detection(sub=None, positions=())
